[Edwin1022] iP

Edwin1022 opened the pull request:
# Friday
> “The best way to predict the future is to create it.” – Peter Drucker ([source](https://www.nu.edu/chancellors-page/december-2016/))

Friday is designed to help you manage your tasks efficiently. With Friday, you can:

* Add tasks quickly
* List all tasks in a clear view
* Mark tasks as done or unmark them
* Delete tasks that you no longer need

How to Use?

1. Download the latest version from [GitHub Releases](https://github.com/Edwin1022/ip/releases).
2. Install the chatbot.
3. Add your tasks.
4. 🎉 Let Friday take care of your tasks so you can focus on what matters!

And it is **FREE**!

Features

- [x] Task management
- [ ] Deadline management (coming soon)
- [ ] Reminders (coming soon)
- [x] Add date and time handling

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Edwin1022 committed:
Rename chatbot

Edwin1022 committed:
Add greeting

Edwin1022 committed:
Add support for echo

Edwin1022 committed:
Add support for addTask and listTasks

Edwin1022 committed:
Add Task class

Edwin1022 committed:
Add support for mark and unmark tasks as done

Edwin1022 committed:
Add Todo class

Edwin1022 committed:
Add Deadline class

Edwin1022 committed:
Add Event class

Edwin1022 committed:
Add support for tracking ToDos, Deadlines and Events

Edwin1022 committed:
Add Automated Text UI Testing

Edwin1022 committed:
Add support for error handling

Edwin1022 committed:
Add support for deleteTask

Edwin1022 committed:
Add Command interface

Edwin1022 committed:
Add Parser class

Edwin1022 committed:
Enchance sustainability of the code

Edwin1022 committed:
Add Storage class

Edwin1022 committed:
Add support for saving the tasks in the hard disk and loading the tasks from the hard disk

Edwin1022 committed:
Merge branch 'branch-Level-7'

Edwin1022 committed:
Add support for understanding dates and times

Edwin1022 committed:
Merge branch 'branch-Level-8'

Edwin1022 committed:
Merge branch 'add-gradle-support'

Edwin1022 committed:
Automate project builds using Gradle

Edwin1022 committed:
Add JUnit tests

Edwin1022 committed:
Package the App as a JAR file

Edwin1022 committed:
Add JavaDoc comments

Edwin1022 committed:
Revert "Add JavaDoc comments"

This reverts commit c9110d21fc40617e2cf3a2f596610dc4438bd967.

Edwin1022 committed:
Add JavaDoc comments

Edwin1022 committed:
Modify JavaDoc comments

Edwin1022 committed:
Modify the ordering of import statements

Edwin1022 committed:
Put Parser, Storage, TaskList and Ui classes into util package

Edwin1022 committed:
Remove unnecessary use of this with fields

Edwin1022 committed:
Improve layout of the code base

Edwin1022 committed:
Add support for finding a task

Edwin1022 committed:
Merge branch 'branch-A-JavaDoc'

Edwin1022 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/friday/task/Deadline.java
#	src/main/java/friday/task/Event.java
#	src/main/java/friday/task/Task.java
#	src/main/java/friday/util/Parser.java
#	src/test/java/friday/TaskListTest.java

Edwin1022 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/friday/TaskList.java
#	src/main/java/friday/util/Parser.java
#	src/main/java/friday/util/Ui.java

Edwin1022 committed:
Add JavaDoc comments for increment Level-9

Edwin1022 committed:
Modify support for understanding dates and times

Edwin1022 committed:
Use CheckStyle

Edwin1022 committed:
Merge branch 'branch-A-CheckStyle'

Edwin1022 committed:
Create GUI for Friday

starchypotatocode commented on src/main/java/friday/Friday.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public void run() {
+        ui.greet();
+        boolean isExit = false;
```
Would shouldExit be a more suitable name, and also fulfill the naming criteria?

starchypotatocode commented on src/main/java/friday/task/Task.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Marks the Task as done.
+     */
+    public void markAsDone() {
```
Should this be setDone instead to fulfill the method naming criteria for setters?

starchypotatocode commented on src/main/java/friday/task/Task.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Unmarks the Task, marking it as not done.
+     */
+    public void unmarkAsDone() {
```
Likewise this setNotDone?

starchypotatocode commented on src/main/java/friday/util/Storage.java:
```diff
@@ -0,0 +1,88 @@
+            }
+            scanner.close();
+        } else {
+            file.getParentFile().mkdirs();
```
Be careful of situations where file might be null

starchypotatocode commented on src/main/java/friday/util/Ui.java:
```diff
@@ -0,0 +1,183 @@
+     */
+    public void greet() {
+        System.out.println(
+                "    ____________________________________________________________\n"
```
Might be better to factor these lines out as a separate function?

starchypotatocode commented on src/test/java/friday/TaskListTest.java:
```diff
@@ -0,0 +1,55 @@
+     * Verifies that the task is correctly removed from the list and that the size of the list is updated.
+     */
+    @Test
+    public void testDeleteTaskWithValidIndex() {
```
This works, but you could also try something like deleteTask_validIndex_success?

keithxun commented on src/main/java/friday/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+ * Represents an abstract command in the command pattern.
+ * All concrete command classes will inherit from this class and implement specific command functionality.
+ */
+public abstract class Command {
```
I like how you use an abstract command class to manage commands

keithxun commented on src/main/java/friday/command/OnCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * Represents a command to list tasks occurring on a specific date.
+ * Inherits from the Command class and provides functionality to filter and display tasks based on a given date.
+ */
+public class OnCommand extends Command {
```
Is there a better name than OnCommand? Its function is not obvious by name

keithxun commented on src/main/java/friday/task/Event.java:
```diff
@@ -0,0 +1,60 @@
+    public Event(String description, String from, String to, boolean isDone) {
+        super(description, isDone);
+        this.from = LocalDateTime.parse(from, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+        this.to = LocalDateTime.parse(to, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
```
I like how you can handle time too!

keithxun commented on src/main/java/friday/util/Parser.java:
```diff
@@ -0,0 +1,196 @@
+import friday.command.UnmarkCommand;
+import friday.task.Deadline;
+import friday.task.Event;
+import friday.task.Todo;
```
```suggestion
import friday.task.*
```

Would it be better to import all since you are using all of them anyways?

starchypotatocode replied:
No, it contradicts the style guide so you shouldn't do that?

keithxun reviewed (commented):
Other than some naming suggestions, lgtm

Edwin1022 committed:
Improve overall code quality

Edwin1022 committed:
Merge branch 'branch-Level-10'

Edwin1022 committed:
Use assert feature in the code

Edwin1022 committed:
Use Streams feature to filter tasks

Edwin1022 committed:
Improve TaskList class by eliminating the use of null to avoid
NullPointerException

The TaskList class methods return null for invalid indices, which can
lead to NullPointerExceptions when the returned value is used without
proper checks.

Returning null for invalid indices is error-prone and can cause runtime
exceptions if not handled correctly.

Let's update the TaskList class methods to throw an appropriate
exception instead of returning null for invalid indices. This change
ensures that invalid operations are caught early and handled explicitly,
improving the robustness of the code.

Using exceptions for error handling is preferable as it makes the code
more predictable and easier to debug. It also enforces proper error
handling by the caller.

No other relevant information.

Edwin1022 committed:
Delete friday.txt

Edwin1022 committed:
Use Streams feature to find tasks by date

Edwin1022 committed:
Improve overall code readability

Currently, the code contains complicated expressions with many negations
and nested parentheses, magic literals, multiple levels of abstraction
within a single code fragment, and the happy path is not prominent.

This needs to change because such practices make the code harder to
read, understand, and maintain. Complicated expressions and nested
parentheses can confuse developers, magic literals reduce code clarity,
multiple levels of abstraction within a single fragment can make the
code less modular, and not having the happy path prominent can make the
code flow less intuitive.

Refactor the code to:
- Simplify expressions by breaking down complex conditions and reducing
nested parentheses.
- Replace magic literals with named constants to improve clarity.
- Ensure each code fragment maintains a single level of abstraction.
- Make the happy path prominent by restructuring the code flow.

These changes are done to enhance code readability, maintainability,
and overall quality, making it easier for developers to understand and
work with the codebase.

No other changes are made to the functionality of the code.

Edwin1022 committed:
Merge pull request #1 from Edwin1022/branch-A-Assertions

Merge branch 'branch-A-Assertions'

Edwin1022 committed:
Merge branch 'master' of https://github.com/Edwin1022/ip

Edwin1022 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/friday/Friday.java
#	src/main/java/friday/command/DeleteCommand.java
#	src/main/java/friday/command/MarkCommand.java
#	src/main/java/friday/command/UnmarkCommand.java
#	src/main/java/friday/task/Deadline.java
#	src/main/java/friday/task/Event.java
#	src/main/java/friday/util/Parser.java
#	src/main/java/friday/util/Storage.java
#	src/main/java/friday/util/TaskList.java
#	src/main/java/friday/util/Ui.java

Edwin1022 committed:
Merge pull request #2 from Edwin1022/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality'

Edwin1022 committed:
Merge branch 'master' of https://github.com/Edwin1022/ip

Edwin1022 committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/friday/command/FindCommand.java
#	src/main/java/friday/util/TaskList.java

Edwin1022 committed:
Merge pull request #3 from Edwin1022/branch-A-Streams

Merge branch 'branch-A-Streams'

Edwin1022 committed:
Merge branch 'master' of https://github.com/Edwin1022/ip

Edwin1022 committed:
Do some refactoring

Edwin1022 committed:
Set up Continuous Integration (CI)

Edwin1022 committed:
Add support to sort deadlines and events chronologically

Edwin1022 committed:
Merge branch 'branch-BCD-Extension'

Edwin1022 committed:
Improve user experience

Edwin1022 committed:
Improve the GUI

Edwin1022 committed:
Give a unique personality to the chatbot

Edwin1022 committed:
Merge branch 'branch-A-BetterGui'

Edwin1022 committed:
Merge branch 'branch-A-Personality'

# Conflicts:
#	src/main/java/friday/util/Ui.java

Edwin1022 committed:
Handle date error

Edwin1022 committed:
Rename .yml file

Edwin1022 committed:
Test all the methods in Storage class and TaskList class

Edwin1022 committed:
Merge branch 'branch-A-MoreErrorHandling'

Edwin1022 committed:
Merge branch 'branch-A-MoreTesting'

Edwin1022 committed:
Add a User Guide to the project

Edwin1022 committed:
Merge branch 'branch-A-UserGuide'
